speechbrain/utils/check_yaml.py

```diff
@@ -0,0 +1,112 @@
+        included).
+    """
+    var_lst = []
+    with open(hparam_file) as f:
+        for line in f:
+            # Avoid empty lists or comments
+            if len(line) <= 1 or line.startswith("#"):
+                continue
+            # Remove trailing characters
+            line = line.rstrip()
+
+            # Check for variables (e.g., key:)
+            if line.find(":") != -1:
+                var_name = line[: line.find(":")]
+                # The variables to check are like "key:" (we do not need to check
+                # subvariavles as " key:")
+                if not (
+                    var_name[0] == " "
+                    or var_name[0] == "\t"
+                    or "!!" in line
+                    or "!apply" in line
+                ):
+                    var_lst.append(var_name)
+                # Check for the reference pattern
+                if line.find("!ref <") != -1:
+                    sub_var = line[line.find("!ref <") + 6 : line.find(">")]
+                    # Remove variables already used in yaml
+                    if sub_var in var_lst:
+                        var_lst.remove(sub_var)
+    return var_lst
+
+
+def check_yaml_vs_script(hparam_file, script_file):
+    """Checks consistency between the given yaml file (hparams_file) and the
+    script file. The function detects if there are variables declared in the yaml
+    file, but not used in the script file.
+
+    Arguments
+    ---------
+    hparam_file : path
+        Path of the yaml file containing the hyperparameters.
+    script_file : path
+        Path of the script file (.py) containing the training recipe.
+
+    Returns
+    -------
+    Bool
+        This function returns False if some mismatch is detected and True otherwise.
+        A warning is raised to inform about which variable has been declared but
+        not used.
+    """
+    print("Checking %s..." % (hparam_file))
+
+    # Check if files exist
+    if not (os.path.exists(hparam_file)):
+        print("File %s not found!" % (hparam_file,))
+        return False
+
+    if not (os.path.exists(script_file)):
+        print("File %s not found!" % (script_file,))
+        return False
+
+    # Getting list of variables declared in yaml
+    var_lst = get_yaml_var(hparam_file)
+
+    # Detect which of these variables are used in the script file
+    detected_var = []
+    with open(script_file) as f:
+        for line in f:
+            for var in var_lst:
+                # The pattern can be ["key"] or ".key"
+                if '["' + var + '"]' in line:
+                    if var not in detected_var:
+                        detected_var.append(var)
+                if "." + var in line:
+                    if var not in detected_var:
+                        detected_var.append(var)
+
+    # Check which variables are declared but not used
+    unused_vars = list(set(var_lst) - set(detected_var))
```

Comment:
This filters for the hparam keys declared but unused in train.py - how about the keys asked for in train.py but not defined in hparams? It's just for an overview of which default variables are used - so no error needs to be thrown.

`default_vars = list(set(detected_var) - set(var_lst))`

// just a thought on exhaustivity

Reply:
I think we can potentially add this check as well. However, If train.py uses a variable not declared in the yaml file, there will be an error at runtime. So, we can still detect these errors with the recipe tests that we will hopefully setup. As far as I remember, some recipes define new hparams (like `self.hparams.new_varable=value`) in the train script directly. This is not a good practice, but it was needed for some reason I don't recall. It might be a bit trickier to set up this check.

Reply:
Should the `self.hparams.new_varable=value` bad practice be checked for?

btw (untested)
```python
default_run_opt_keys = ["debug", "debug_batches", "debug_epochs", "device": "cpu", "data_parallel_backend", "distributed_launch", "distributed_backend", "find_unused_parameters", "jit_module_keys", "auto_mix_prec", "max_grad_norm", "nonfinite_patience", "noprogressbar", "ckpt_interval_minutes", "grad_accumulation_factor", "optimizer_step_limit"]
unused_vars = list(set(var_lst) - set(detected_var) - set(default_run_opt_keys))
```

Reply:
The `progressbar` logic causes 22x warnings - while it is in hparams as `"noprogressbar"`

https://github.com/speechbrain/speechbrain/blob/d6736adaabf0c6a1021dd1970056f9416bcdcd62/speechbrain/core.py#L1022

```python
        if progressbar is None:
            progressbar = not self.noprogressbar
```

=> simply add "progressbar" to the above default list?